jmclient/jmclient/wallet.py

```diff
@@ -1292,6 +1305,26 @@ def sign_psbt(self, in_psbt, with_sign_result=False):
                     #key is ((privkey, locktime), engine) for timelocked addrs
                     key = (key[0][0], key[1])
                 privkeys.append(key)
+
+        # in the rare situation that we want to sign a psbt using private keys
```

Comment:
I understand this addition, it's well conceived.

But I am concerned about the use case here. One of the main ways that PSBTs protect a user is by having all the inputs of a transaction referenced, so the "user" (usually in code) can check that they are only spending what they expect to spend.

That's of course why there's a big scary warning at the top of this `sign_psbt` function, it can and should *only* be called by callers who already are confident that they know the source of each input.

For the use case in `taker_utils.direct_send()` this is fine: our wallet is acting as Creator in the PSBT roles, and so there is no danger of confusion over inputs.

In the case of payjoin, there are detailed checks in the `PayjoinManager` to ensure that rules are followed.

So I think we're OK with this, it's maybe the case that we should alter the logic of what `sign_psbt` does to avoid nasty accidents in future feature additions.

Reply:
I already have a `bumpfee` script that I intend to add to the repository to handle bumping RBF transactions if this gets merged. I remember figuring out that when the wallet signs a transaction, it removes any "spent" utxos from the utxo store and the only way it's able to access available private keys to use in signing future transactions, is through the utxo store (at least in the PSBT case). For RBF transactions, the private key needs to be accessible to sign future replacements and utxos are not "spent" until they're confirmed. This was included as a way to lookup the private key even for utxos that were initially deemed spent. The only side effect I see is adding the ability to sign an input that is already spent and having the transaction rejected because it's a double spend.

Reply:
Yes that's a very good point. Spent utxos are not considered available, and that applies across restarts of course.

There's probably different ways you could go, in the task of rbf bumping the transaction, by just looking up the transaction, making an object, and calling `BaseWallet.sign_transaction` which will do a path-> privkey conversion before calling `jmbitcoin.sign` on the indices. I don't think that that would be affected by the contents of the `UTXOManager`, would it?